src/compiler/parser.ts

```diff
@@ -4665,10 +4665,13 @@ namespace ts {
             const propertyAccess = <PropertyAccessExpression>createNode(SyntaxKind.PropertyAccessExpression, expression.pos);
             propertyAccess.expression = expression;
             propertyAccess.questionDotToken = questionDotToken;
-            // checker will error on private identifiers in optional chains, so don't have to catch them here
             propertyAccess.name = parseRightSideOfDot(/*allowIdentifierNames*/ true, /*allowPrivateIdentifiers*/ true);
             if (questionDotToken || expression.flags & NodeFlags.OptionalChain) {
                 propertyAccess.flags |= NodeFlags.OptionalChain;
+                if (isPrivateIdentifier(propertyAccess.name)) {
+                    parseErrorAtRange(propertyAccess.name, Diagnostics.An_optional_chain_cannot_contain_private_identifiers);
+                    propertyAccess.name = createMissingNode(SyntaxKind.Identifier, /*reportAtCurrentPosition*/ false);

```

Comment:
I think it would be better to just put the privatename identifier on the node so that the language service still works

Reply:
Should this be a grammar error in the binder @sandersn?

Reply:
I was looking around and it seems like diagnostics from the binder and checker are discarded for JS files that have checkJs off.

https://github.com/microsoft/TypeScript/blob/18269c0fc8a9dcccc8f9061e0c88f34f1135e1ca/src/compiler/program.ts#L1736-L1737